File: libs/engine/src/accounts/signing.cc
#include "accounts/signing.h"

#include <stdexcept>

namespace hestia::engine {
    namespace {
        constexpr char kStandardAlphabet[] = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
        constexpr char kUrlAlphabet[] = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_";

        std::string base64_encode(const std::vector<std::uint8_t> &data, const char *alphabet, bool pad) {
            std::string out;
            out.reserve((data.size() + 2) / 3 * 4);
            std::size_t i = 0;
            for (; i + 3 <= data.size(); i += 3) {
                const std::uint32_t n = (data[i] << 16) | (data[i + 1] << 8) | data[i + 2];
                out.push_back(alphabet[(n >> 18) & 0x3F]);
                out.push_back(alphabet[(n >> 12) & 0x3F]);
                out.push_back(alphabet[(n >> 6) & 0x3F]);
                out.push_back(alphabet[n & 0x3F]);
            }
            if (const std::size_t rest = data.size() - i; rest == 1) {
                const std::uint32_t n = data[i] << 16;
                out.push_back(alphabet[(n >> 18) & 0x3F]);
                out.push_back(alphabet[(n >> 12) & 0x3F]);
                if (pad) out.append("==");
            } else if (rest == 2) {
                const std::uint32_t n = (data[i] << 16) | (data[i + 1] << 8);
                out.push_back(alphabet[(n >> 18) & 0x3F]);
                out.push_back(alphabet[(n >> 12) & 0x3F]);
                out.push_back(alphabet[(n >> 6) & 0x3F]);
                if (pad) out.push_back('=');
            }
            return out;
        }

        void append_be32(std::vector<std::uint8_t> &buf, std::uint32_t value) {
            buf.push_back(static_cast<std::uint8_t>(value >> 24));
            buf.push_back(static_cast<std::uint8_t>(value >> 16));
            buf.push_back(static_cast<std::uint8_t>(value >> 8));
            buf.push_back(static_cast<std::uint8_t>(value));
        }

        void append_be64(std::vector<std::uint8_t> &buf, std::uint64_t value) {
            for (int shift = 56; shift >= 0; shift -= 8) {
                buf.push_back(static_cast<std::uint8_t>(value >> shift));
            }
        }

        void append_bytes(std::vector<std::uint8_t> &buf, const std::string &value) {
            buf.insert(buf.end(), value.begin(), value.end());
        }
    } // namespace

    std::string format_uuid_v4(std::vector<std::uint8_t> bytes) {
        if (bytes.size() != 16) {
            throw std::runtime_error("a UUID needs exactly 16 bytes");
        }
        bytes[6] = static_cast<std::uint8_t>((bytes[6] & 0x0F) | 0x40);
        bytes[8] = static_cast<std::uint8_t>((bytes[8] & 0x3F) | 0x80);
        constexpr char hex[] = "0123456789abcdef";
        std::string out;
        out.reserve(36);
        for (std::size_t i = 0; i < bytes.size(); ++i) {
            if (i == 4 || i == 6 || i == 8 || i == 10) out.push_back('-');
            out.push_back(hex[bytes[i] >> 4]);
            out.push_back(hex[bytes[i] & 0x0F]);
        }
        return out;
    }

    std::string base64_standard(const std::vector<std::uint8_t> &data) {
        return base64_encode(data, kStandardAlphabet, true);
    }

    std::string base64url_nopad(const std::vector<std::uint8_t> &data) {
        return base64_encode(data, kUrlAlphabet, false);
    }

    std::string xbox_signature_header(const ProofKey &key, const std::string &url_path,
                                      const std::string &authorization, const std::string &body,
                                      std::int64_t unix_time) {
        const auto filetime = static_cast<std::uint64_t>(unix_time + 11644473600LL) * 10000000ULL;

        std::vector<std::uint8_t> message;
        append_be32(message, 1);
        message.push_back(0);
        append_be64(message, filetime);
        message.push_back(0);
        append_bytes(message, "POST");
        message.push_back(0);
        append_bytes(message, url_path);
        message.push_back(0);
        append_bytes(message, authorization);
        message.push_back(0);
        append_bytes(message, body);
        message.push_back(0);

        const auto signature = key.sign(message);
        if (signature.size() != 64) {
            throw std::runtime_error("ECDSA signature was not 64 bytes");
        }

        std::vector<std::uint8_t> envelope;
        append_be32(envelope, 1);
        append_be64(envelope, filetime);
        envelope.insert(envelope.end(), signature.begin(), signature.end());
        return base64_standard(envelope);
    }
} // namespace hestia::engine

File: crates/engine/src/accounts/signing.rs
//! Xbox request signing: an ECDSA P-256 proof key and the FILETIME-stamped
//! `Signature` header. One cross-platform implementation (`p256`) — the C++ tree
//! needed an OpenSSL/CNG split; Rust does not.

use base64::engine::general_purpose::{STANDARD, URL_SAFE_NO_PAD};
use base64::Engine as _;
use p256::ecdsa::signature::Signer;
use p256::ecdsa::{Signature, SigningKey};
use rand_core::{OsRng, RngCore};

/// A per-login ECDSA P-256 proof key. Its public point (x, y) is advertised in
/// the Xbox proof-of-possession JWK; `sign` produces the raw 64-byte (r‖s)
/// signature the `Signature` header carries.
pub struct ProofKey {
    signing_key: SigningKey,
    id: String,
    x: String,
    y: String,
}

impl ProofKey {
    pub fn generate() -> ProofKey {
        let signing_key = SigningKey::random(&mut OsRng);
        let point = signing_key.verifying_key().to_encoded_point(false);
        let x = point.x().expect("P-256 public point has an x coordinate");
        let y = point.y().expect("P-256 public point has a y coordinate");
        ProofKey {
            id: format_uuid_v4(&random_bytes(16)),
            x: base64url_nopad(x),
            y: base64url_nopad(y),
            signing_key,
        }
    }

    pub fn id(&self) -> &str {
        &self.id
    }
    pub fn x(&self) -> &str {
        &self.x
    }
    pub fn y(&self) -> &str {
        &self.y
    }

    /// ECDSA/SHA-256 over `message`, serialized as fixed-width raw r‖s (64 bytes).
    /// Xbox verifies against the advertised proof key, so a fresh (randomized or
    /// deterministic) signature is accepted — byte-identity with any other signer
    /// is neither required nor possible.
    pub fn sign(&self, message: &[u8]) -> Vec<u8> {
        let signature: Signature = self.signing_key.sign(message);
        signature.to_bytes().to_vec()
    }
}

pub fn random_bytes(count: usize) -> Vec<u8> {
    let mut out = vec![0u8; count];
    OsRng.fill_bytes(&mut out);
    out
}

pub fn format_uuid_v4(bytes: &[u8]) -> String {
    assert_eq!(bytes.len(), 16, "a UUID needs exactly 16 bytes");
    let mut b = [0u8; 16];
    b.copy_from_slice(bytes);
    b[6] = (b[6] & 0x0F) | 0x40; // version 4
    b[8] = (b[8] & 0x3F) | 0x80; // variant
    let mut out = String::with_capacity(36);
    for (i, byte) in b.iter().enumerate() {
        if matches!(i, 4 | 6 | 8 | 10) {
            out.push('-');
        }
        out.push_str(&format!("{byte:02x}"));
    }
    out
}

pub fn base64_standard(data: &[u8]) -> String {
    STANDARD.encode(data)
}

pub fn base64url_nopad(data: &[u8]) -> String {
    URL_SAFE_NO_PAD.encode(data)
}

pub fn hex(bytes: &[u8]) -> String {
    let mut s = String::with_capacity(bytes.len() * 2);
    for b in bytes {
        s.push_str(&format!("{b:02x}"));
    }
    s
}

/// Build the base64 `Signature` header: a length-delimited message
/// (version, FILETIME, method, path, authorization, body) is signed, then the
/// (version ‖ FILETIME ‖ signature) envelope is base64-encoded.
pub fn xbox_signature_header(
    key: &ProofKey,
    url_path: &str,
    authorization: &str,
    body: &str,
    unix_time: i64,
) -> String {
    let filetime = (unix_time as u64).wrapping_add(11_644_473_600) * 10_000_000;

    let mut message = Vec::new();
    message.extend_from_slice(&1u32.to_be_bytes());
    message.push(0);
    message.extend_from_slice(&filetime.to_be_bytes());
    message.push(0);
    message.extend_from_slice(b"POST");
    message.push(0);
    message.extend_from_slice(url_path.as_bytes());
    message.push(0);
    message.extend_from_slice(authorization.as_bytes());
    message.push(0);
    message.extend_from_slice(body.as_bytes());
    message.push(0);

    let signature = key.sign(&message);
    debug_assert_eq!(signature.len(), 64, "ECDSA signature must be 64 bytes");

    let mut envelope = Vec::with_capacity(4 + 8 + signature.len());
    envelope.extend_from_slice(&1u32.to_be_bytes());
    envelope.extend_from_slice(&filetime.to_be_bytes());
    envelope.extend_from_slice(&signature);
    base64_standard(&envelope)
}

#[cfg(test)]
mod tests {
    //! Signing oracle tests. The C++ tree was the byte-diff oracle before the
    //! cutover; with one cross-platform `p256` signer there is nothing to diff
    //! against, so these pin the wire-visible envelope layout and lean on ECDSA's
    //! RFC 6979 determinism to catch a nonce/serialization regression.

    use super::*;
    use p256::ecdsa::signature::Verifier;
    use p256::ecdsa::{Signature, SigningKey, VerifyingKey};

    impl ProofKey {
        /// Build a proof key from a fixed scalar so a test signs deterministically.
        fn from_scalar_bytes(bytes: &[u8; 32]) -> ProofKey {
            let signing_key = SigningKey::from_slice(bytes).expect("valid P-256 scalar");
            let point = signing_key.verifying_key().to_encoded_point(false);
            let x = point.x().expect("x coordinate");
            let y = point.y().expect("y coordinate");
            ProofKey {
                id: "00000000-0000-4000-8000-000000000000".to_string(),
                x: base64url_nopad(x),
                y: base64url_nopad(y),
                signing_key,
            }
        }

        fn verifying_key(&self) -> VerifyingKey {
            *self.signing_key.verifying_key()
        }
    }

    /// Rebuild the exact message `xbox_signature_header` signs, so the test can
    /// verify the extracted signature against it.
    fn signed_message(filetime: u64, path: &str, authorization: &str, body: &str) -> Vec<u8> {
        let mut message = Vec::new();
        message.extend_from_slice(&1u32.to_be_bytes());
        message.push(0);
        message.extend_from_slice(&filetime.to_be_bytes());
        message.push(0);
        message.extend_from_slice(b"POST");
        message.push(0);
        message.extend_from_slice(path.as_bytes());
        message.push(0);
        message.extend_from_slice(authorization.as_bytes());
        message.push(0);
        message.extend_from_slice(body.as_bytes());
        message.push(0);
        message
    }

    #[test]
    fn filetime_conversion_is_a_fixed_vector() {
        // The Windows FILETIME epoch is 1601-01-01; 11_644_473_600 s before the
        // Unix epoch, in 100 ns ticks.
        let key = ProofKey::from_scalar_bytes(&[0x42; 32]);
        let header = xbox_signature_header(&key, "/authorize", "", "{}", 0);
        let envelope = STANDARD.decode(header).expect("base64 envelope");
        let filetime = u64::from_be_bytes(envelope[4..12].try_into().unwrap());
        assert_eq!(filetime, 116_444_736_000_000_000);
    }

    #[test]
    fn signature_header_envelope_layout() {
        let key = ProofKey::from_scalar_bytes(&[0x11; 32]);
        let header = xbox_signature_header(
            &key,
            "/device/authenticate",
            "XBL3.0 x=1;t",
            "{}",
            1_700_000_000,
        );
        let envelope = STANDARD.decode(header).expect("base64 envelope");
        // version(4) + filetime(8) + raw r‖s signature(64)
        assert_eq!(envelope.len(), 4 + 8 + 64);
        assert_eq!(u32::from_be_bytes(envelope[0..4].try_into().unwrap()), 1);
    }

    #[test]
    fn signature_verifies_against_the_proof_key() {
        let key = ProofKey::from_scalar_bytes(&[0x11; 32]);
        let (path, auth, body, unix) =
            ("/authorize", "XBL3.0 x=1;t", "{\"a\":1}", 1_700_000_000i64);
        let header = xbox_signature_header(&key, path, auth, body, unix);
        let envelope = STANDARD.decode(header).expect("base64 envelope");

        let filetime = (unix as u64 + 11_644_473_600) * 10_000_000;
        let message = signed_message(filetime, path, auth, body);
        let signature = Signature::from_slice(&envelope[12..]).expect("64-byte signature");
        key.verifying_key()
            .verify(&message, &signature)
            .expect("the envelope's signature must verify against the advertised proof key");
    }

    #[test]
    fn signing_is_deterministic() {
        // RFC 6979 nonces make two keys built from the same scalar sign a message
        // identically; a regression to randomized nonces would break replay diffs.
        let a = ProofKey::from_scalar_bytes(&[0x24; 32]);
        let b = ProofKey::from_scalar_bytes(&[0x24; 32]);
        let message = b"hestia-signing-oracle";
        assert_eq!(a.sign(message), b.sign(message));
        assert_eq!(a.sign(message), a.sign(message));
    }
}
